src/hotspot/share/utilities/unalignedAccess.hpp

```diff
@@ -0,0 +1,187 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_UNALIGNED_ACCESS_HPP
+#define SHARE_UTILITIES_UNALIGNED_ACCESS_HPP
+
+#include "memory/allStatic.hpp"
+#include "metaprogramming/enableIf.hpp"
+#include "metaprogramming/primitiveConversions.hpp"
+#include "utilities/globalDefinitions.hpp"
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <type_traits>
+
+#if defined(ADDRESS_SANITIZER)
+#include <sanitizer/common_interface_defs.h>
+#endif
+
+template <typename T, size_t S = sizeof(T), size_t A = alignof(T)>
+struct UnalignedLoadImpl;
+template <typename T, size_t S = sizeof(T), size_t A = alignof(T)>
+struct UnalignedStoreImpl;
+
+// Support for well defined potentially unaligned memory access, regardless of underlying
+// architecture support.
+//
+// Unaligned access is undefined behavior according to the standard. Some architectures support
+// aligned and unaligned memory access via the same instructions (i.e. x86, AArch64) while some do
+// not permit unaligned access at all. Compilers are free to assume that all memory access of a
+// type T are done at a suitably aligned address for type T, that is an address aligned to
+// alignof(T). This is not always the case, as there are use cases where we may want to access type
+// T at a non-suitably aligned address. For example, when serializing scalar types to a buffer
+// without padding.
+//
+// IMPORTANT: On some architectures the cost for unaligned accesses is cheap, while on others it is
+// expensive. Only use unaligned accesses when necessary.
+class UnalignedAccess final : public AllStatic {
+ public:
+  // Loads the bits of the value of type T from the specified address. The address may or may not be
+  // suitably aligned for type T. T must be trivially copyable and must be default constructible.
+  template <typename T, ENABLE_IF(std::is_trivially_copyable<T>::value &&
+                                  std::is_default_constructible<T>::value)>
+  static inline T load(const void* p) {
+    return UnalignedLoadImpl<T>{}(p);
+  }
+
+  // Stores the bits of the value of type T at the specified address. The address may or may not be
+  // suitably aligned for type T. T must be trivially copyable and must be default constructible.
+  template <typename T, ENABLE_IF(std::is_trivially_copyable<T>::value)>
+  static inline void store(void* p, T x) {
+    UnalignedStoreImpl<T>{}(p, x);
+  }
+};
+
+template <typename T, size_t S, size_t A>
+struct UnalignedLoadImpl {
+  inline T operator()(const void* p) const {
+#if defined(TARGET_COMPILER_gcc) || defined(TARGET_COMPILER_xlc)
+    // When available, explicitly prefer the builtin memcpy variant. This ensures GCC/Clang will
+    // do its best at generating optimal machine code regardless of build options. For architectures
```

Comment:
In what circumstances will GCC do better with  __builtin_memcpy(), and why will we care?

Reply:
If `-O2` is not passed, it will emit calls to libc `memcpy`. Depending on where this is used, that could significantly slow things down. Today lots of places just to raw load and store, which are single instructions, but undefined behavior. I was aiming to have switching be as close to equivalent as possible.

Reply:
Won't it be better to use put_native_u2/4/8 here instead of __builtin_memcpy ? 
So we can control everything in jvm code.